### DISCUSSION TITLE
"Variable 'i' is not accessed" warning in pandas query with '@'

### DISCUSSION THREAD

**Discussion by Pastells** [microsoft/pyright#7492] [Q&A]
When refering to a variable inside a pandas query with '@', pyright doesn't recognize it.

```python
import pandas as pd

df = pd.DataFrame(
    {"id": [1, 1, 1, 2, 2, 2], "pos": [0, 1, 2, 0, 1, 2], "class": [0, 0, 1, 0, 1, 0]}
)

clas_pos = [df.query("id == @i")["class"].to_numpy() for i in df.comment_id]   # Warning: "i" is not accessed
```

**Comment by erictraut** [comment]
This is by design, so not a bug. A static analyzer has no way of knowing that a library is going to use introspection to read a local variable through "magic" means. From the perspective of standard coding practices and static analysis, `i` is not referenced in this code.

**Reply by Pastells** [reply]
That makes sense. Do you think it is a bad practice to use this syntax, though?

**Reply by erictraut** [reply]
I don't know enough about pandas to answer that question. Does pandas provide an alternative mechanism that doesn't involve magic introspection of local variables referenced within string literals? If so, that might be a better solution. In any case, standards-based static analysis tools like pyright will likely never understand this sort of library-specific syntax / semantics.

**Reply by Pastells** [reply]
Yes, the (probably more standard) way would be:

```python
clas_pos = [
    df[df["id"] == i]["class"].to_numpy()
    for i in df.comment_id
]
```

I will stick to that.

Thank you!
